corehq/messaging/tasks.py

```diff
@@ -146,4 +148,17 @@ def run_messaging_rule(domain, rule_id):
     # complete at about the time that the last tasks are finishing up.
     # This beats saving the task results in the database and using a
     # celery chord which would be more taxing on system resources.
-    set_rule_complete.delay(rule_id)
+    queue_task_with_retries(set_rule_complete, rule_id)
+
+
+def queue_task_with_retries(task_to_queue, *args, **kwargs):
+    retries = 0
+    while retries < 5:
+        try:
+            task_to_queue.delay(*args, **kwargs)
+            # break from while if no issues
+            break
+        except Exception:
```

Comment:
This is very broad and seems like it might catch errors that should not be retried. Can it be changed to a more limited exception type? For example, what exception was raised in the scenario that motivated this change?

Reply:
ya that was intentional @millerdev , to catch any exception. The idea is to catch any sort on temporary/intermittent exception that can be caused due to say, infra issues. So catching just the error that was highlighted in sentry was not enough.
Since it's a simple delay call, I don't think it will have any other source of error.
And even then, if there is a genuine error, it would fail after 5 attempts.

Reply:
Do you have evidence that `task.delay()` throws many different types of exceptions? (It sounds like you only know of one so far.) The one that was raised (`SoftTimeLimitExceeded`) seems like a strange error to be raised here. Do you know why it was raised? Would it be reasonable to increase the time limit (maybe to 5 minutes) instead catching/retrying?

While researching `SoftTimeLimitExceeded` I remembered that Celery has its own built-in retry mechanism. Can you use that instead of rolling your own?

My experience is that there are typically a very limited set of errors that should be retried and all others should be handled some other way. More types can be added later if we discover others that should be retried.

Reply:
Thanks for the close review @millerdev 
1. Yup, that is the only exception i have seen so far. I think it happened because RabbitMQ was not accessible and celery waited for sometime before raising that exception.
I am not sure if celery retries this or not because on one instance where the same exception was raised [due to slow/no response from db query](https://sentry.io/organizations/dimagi/issues/1631921255/events/0ac4e33981214f248cf18fa7cf2072db/?project=1545828&statsPeriod=14d), task appeared to resume but on instance when it was raised when queuing another task, the calling task was killed.
2. The intention here is to catch any intermittent error to not lose all the work that has been done so far, which can be days as well. Hence used the open Exception class.
I am happy to just catch `SoftTimeLimitExceeded` but i am worried if something new surfaces that was temporary, we would still lose days of work.

Reply:
There are many things that could happen to cause a very long running process to crash, many of them cannot be recovered in-process. Maybe a new iteration pattern should be adopted to make it resume from where it left off in case of crash? That might be a better way to solve the days of work lost issue.